### PR TITLE
BUG: check return value of `_buffer_format_string`

### PR DESCRIPTION
**Pull request by mattip**
Test and fix for not checking return value from call to `_buffer_format_string`, exposed when
`datetime64` dtype used in a field.

### CODE REVIEW

**Comment by mattip** [comment]
@charris I don't understand what is missing. I added a check at line 269, there is another recursive call at line 227 but the return value is passed back to the outer call.

**Comment by charris** [comment]
Oops, I was referencing the wrong call. However, I notice that offset is passed in as  `Py_ssize_t *`, but everywhere assigned to `int` (old_offset, new_offset). Might want to check that. The other call does indeed return the return value, but I wonder if it should be doing the offset arithmetic first?

File: numpy/core/src/multiarray/buffer.c [hunk 1]
```diff
@@ -266,8 +266,11 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
             }
 
             /* Insert child item */
-            _buffer_format_string(child, str, obj, offset,
+            ret = _buffer_format_string(child, str, obj, offset,
                                   active_byteorder);
+            if (ret != 0) {
+                return ret;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/buffer.c]: I'd prefer `-1` here - this gives the impression that we actually have multiple types of error code, which is not the case

Comment (mattip) [reply numpy/core/src/multiarray/buffer.c]: Fixing

It is a shame we cannot use the sentinels for success and failure that we define in our headers:
```
#define NPY_FAIL 0
#define NPY_SUCCEED 1
```

Comment (charris) [reply numpy/core/src/multiarray/buffer.c]: The numpy versions are non-standard and weird. They are appropriate for bash scripts, which is why they are used in `int main functions`, but a more standard unix convention is negative numbers for errors, in particular, -1 is popular. Even standardized positive error values are returned as their negative and converted to positive for display.  Oddly, the topic is rarely covered in style guides.

Comment (eric-wieser) [reply numpy/core/src/multiarray/buffer.c]: `NPY_FAIL` and `NPY_SUCCEED` are only for use in `PyArg_ParseTuple`, which is determined by the CPython API.

**Comment by charris** [comment]
I fixed the type in the declaration of old_offset and made the return value explicit. LGTM now.

@eric-wieser Good to go?

File: numpy/core/src/multiarray/buffer.c [hunk 2]
```diff
@@ -175,6 +175,14 @@ _is_natively_aligned_at(PyArray_Descr *descr,
     return 1;
 }
 
+/*
+ * Fill in str with an appropriate PEP 3118 format string, based on
+ * descr. For structured dtypes, calls itself recursively. Each call extends
+ * str at offset then updates offset, and uses  descr->byteorder, (and
+ * possibly the byte order in obj) to determine the byte-order char.
+ *
+ * Returns 0 for succcess, -1 for failure
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/buffer.c]: nit: typo

File: numpy/core/src/multiarray/buffer.c [hunk 3]
```diff
@@ -266,8 +274,11 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
             }
 
             /* Insert child item */
-            _buffer_format_string(child, str, obj, offset,
+            ret = _buffer_format_string(child, str, obj, offset,
                                   active_byteorder);
+            if (ret == -1) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/buffer.c]: Typically the test is just `ret < 0` - that's how we do it everywhere else

**Comment by eric-wieser** [comment]
If this is going to be part of a chain of backported commits, I think you save a bunch of work by rebasing this on the branch point

**Comment by charris** [comment]
@eric-wieser backports are generally pretty easy, this one should not be a problem.

**Pull request merged**